### PR TITLE
SP-315 get transaction

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: Library/SPIModels.m [hunk 1]
```diff
@@ -140,12 +143,23 @@ - (void)cancelFailed:(NSString *)msg {
...
 }
 
+- (void)callingGt:(NSString *)gtRequestId {
+    self.isAwaitingGtResponse = YES;
+    self.lastStateRequestTime = [NSDate date];
+    self.gtRequestId = gtRequestId;
+}
+
+
 - (void)callingGlt:(NSString *)gltRequestId {
```
Comment (mattgresham) [inline Library/SPIModels.m:153]: As this won't be used, this one can be removed now

Comment (mattgresham) [inline Library/SPIModels.m]: As this won't be used, this one can be removed now. Also could you remove `gotGltResponse`, `isAwaitingGltResponse`, `gltResponsePosRefId` and `lastGltRequestId ` references too please

File: Library/Model/SPITransaction.m [hunk 2]
```diff
@@ -248,6 +248,133 @@ - (SPIMessage *)toMessage {
...
+    if (cr.length != 0 && mr.length != 0) {
+        NSMutableDictionary *data = self.message.data.mutableCopy;
+        data[@"customer_receipt"] = mr;
+        self.message.data = data.copy;
+    }
+}
+
+@end
+
 @implementation SPIGetLastTransactionResponse : NSObject
```
Comment (mattgresham) [inline Library/Model/SPITransaction.m:378]: Could I please get you to remove the `wasTimeOutOfSyncError` has now that `posCounter` has been implemented this one won't be needed

File: Library/Model/SPITransaction.m [hunk 3]
```diff
@@ -248,6 +248,133 @@ - (SPIMessage *)toMessage {
...
 
 - (instancetype)initWithMessage:(SPIMessage *)message {
```
Comment (mattgresham) [inline Library/Model/SPITransaction.m:380]: Could I also get you to remove the `wasTimeOutOfSyncError` call too, please? Now that `posCounter` is in this is no longer needed

File: Library/SPIClient.m [hunk 4]
```diff
@@ -834,6 +834,50 @@ - (void)initiateSettlementEnquiry:(NSString *)posRefId
...
+                return;
+            }
+            
+            weakSelf.state.flow = SPIFlowTransaction;
+            
+            SPIGetTransactionRequest *gtRequest = [[SPIGetTransactionRequest alloc] initWithPosRefId:posRefId];
+            
+            SPIMessage *gtMessage = [gtRequest toMessage];
+            
+            weakSelf.state.txFlowState = [[SPITransactionFlowState alloc] initWithTid:gtMessage.mid
```
Comment (mattgresham) [inline Library/SPIClient.m]: Should this be `posRefId` as this time around we know the transaction `posRefId` we want?

File: Library/SPIClient.m [hunk 5]
```diff
@@ -834,6 +834,50 @@ - (void)initiateSettlementEnquiry:(NSString *)posRefId
...
+                                                                                 type:SPITransactionTypeGetTransaction
+                                                                          amountCents:0
+                                                                              message:gtMessage
+                                                                                  msg:@"Waiting for EFTPOS connection to make a get transaction request"];
+            
+            [weakSelf.state.txFlowState callingGt:gtMessage.mid];
+            
+            if ([weakSelf send:gtMessage]) {
+                [weakSelf.state.txFlowState sent:@"Asked EFTPOS to get transaction"];
```
Comment (mattgresham) [inline Library/SPIClient.m]: Could you please include the `posRefId` in this message?

File: Library/SPIClient.m [hunk 6]
```diff
@@ -1502,52 +1546,46 @@ - (void)handleErrorEvent:(SPIMessage *)m {
...
-            SPILog(@"received a glt response but we had not asked for one within this transaction. Perhaps leftover from previous one. ignoring.");
+        if (!txState.isAwaitingGtResponse) {
+            SPILog(@"received a gt response but we had not asked for one within this transaction. Perhaps leftover from previous one. ignoring.");
             return;
         }
         
-        if (txState.lastGltRequestId != m.mid) {
-            SPILog(@"received a glt response but the message id does not match the glt request that we sent. strange. ignoring.");
+        if (txState.gtRequestId != m.mid) {
+            SPILog(@"received a gt response but the message id does not match the glt request that we sent. strange. ignoring.");
```
Comment (mattgresham) [inline Library/SPIClient.m]: Small typo
```suggestion
            SPILog(@"received a gt response but the message id does not match the gt request that we sent. strange. ignoring.");
```

File: Library/SPIClient.m [hunk 7]
```diff
@@ -1502,52 +1546,46 @@ - (void)handleErrorEvent:(SPIMessage *)m {
...
+                
+                if (!txState.isAwaitingSignatureCheck) {
+                    SPILog(@"GTR-01: Eftpos is waiting for us to send it signature accept/decline, but we were not aware of this. The user can only really decline at this stage as there is no receipt to print for signing.");
                     
                     SPISignatureRequired *req = [[SPISignatureRequired alloc] initWithPosRefId:txState.posRefId
                                                                                      requestId:m.mid
                                                                                  receiptToSign:@"MISSING RECEIPT\n DECLINE AND TRY AGAIN."];
                     [self.state.txFlowState signatureRequired:req msg:@"Recovered in signature required but we don't have receipt. You may decline then retry."];
-                } else if ([gltResponse isWaitingForAuthCode] && !txState.isAwaitingPhoneForAuth) {
+                } else if ([gtResponse isWaitingForAuthCode] && !txState.isAwaitingPhoneForAuth) {
```
Comment (mattgresham) [inline Library/SPIClient.m]: Looks like this section needs to be moved up one level - https://github.com/mx51/spi-client-js/blob/master/src/Spi.js#L1533-L1539

File: Library/SPIClient.m [hunk 8]
```diff
@@ -1557,84 +1595,91 @@ - (void)handleGetLastTransactionResponse:(SPIMessage *)m {
...
+                    SPILog(@"Waiting for Signature response ... stay waiting.");
                     // No need to publish txFlowStateChanged. Can return;
                     return;
                 }
-            } else if ([gltResponse wasTimeOutOfSyncError]) {
-                // Let's not give up based on a TOOS error.
-                // Let's log it, and ignore it.
-                SPILog(@"Time-Out-Of-Sync error in Get Last Transaction response. Let's ignore it and we'll try again.");
-                // No need to publish txFlowStateChanged. Can return;
+            } else if ([gtResponse isWaitingForAuthCode] && ![txState isAwaitingPhoneForAuth]) {
```
Comment (mattgresham) [inline Library/SPIClient.m:1662]: Then this section becomes - https://github.com/mx51/spi-client-js/blob/master/src/Spi.js#L1540-L1544

File: Library/Model/SPITransaction.m [hunk 9]
```diff
@@ -248,6 +248,133 @@ - (SPIMessage *)toMessage {
...
+
+- (BOOL)isWaitingForSignatureResponse {
+    return [self.message.error hasPrefix:@"OPERATION_IN_PROGRESS_AWAITING_SIGNATURE"];
+}
+
+- (BOOL)isWaitingForAuthCode {
+    return [self.message.error hasPrefix:@"OPERATION_IN_PROGRESS_AWAITING_PHONE_AUTH_CODE"];
+}
+
+- (BOOL)IsSomethingElseBlocking {
```
Comment (mattgresham) [inline Library/Model/SPITransaction.m]: Sorry, small nitpick, can you please change the case of this one?
```suggestion
- (BOOL)isSomethingElseBlocking {
```

File: Library/SPIClient.m [hunk 10]
```diff
@@ -1557,84 +1595,91 @@ - (void)handleGetLastTransactionResponse:(SPIMessage *)m {
...
+                [txState completed:SPIMessageSuccessStateFailed response:m msg:[NSString stringWithFormat:@"PosRefId invalid for %@", txState.posRefId]];
+            } else if ([gtResponse isMissingArgumentsError]) {
+                SPILog(@"GTR-06: Get transaction failed, PosRefId is missing.");
+                [txState completed:SPIMessageSuccessStateFailed response:m msg:[NSString stringWithFormat:@"PosRefId is missing for %@", txState.posRefId]];
+            } else if ([gtResponse IsSomethingElseBlocking]) {
+                SPILog(@"GTR-07: Terminal is Blocked by something else... stay waiting.");
                 return;
+            } else {
+                // get transaction failed, but we weren't given a specific reason
+                SPILog(@"GTR-08: Unexpected Response in Get Transaction - Received posRefId:%@ %@", txState.posRefId, m.error);
```
Comment (mattgresham) [inline Library/SPIClient.m]: Small change here
```suggestion
                SPILog(@"GTR-08: Unexpected Response in Get Transaction - Received posRefId:%@ Error:%@", txState.posRefId, m.error);
```

File: Library/SPIClient.m [hunk 11]
```diff
@@ -1557,84 +1595,91 @@ - (void)handleGetLastTransactionResponse:(SPIMessage *)m {
...
+            SPILog(@"Received glt response but we were not expecting one. ignoring.");
+            return;
+        }
+        
+        SPILog(@"Got Last Transaction Response.");
+        
+        SPIGetLastTransactionResponse *gltResponse = [[SPIGetLastTransactionResponse alloc] initWithMessage:m];
+        
+        if (!gltResponse.wasRetrievedSuccessfully) {
+            SPILog(@"Error in Response for Get Last Transaction - Received posRefId:%@ %@. UnknownCompleted.", [gltResponse getPosRefId], [m error]);
```
Comment (mattgresham) [inline Library/SPIClient.m]: I'll get you to update this one too, please
```suggestion
            SPILog(@"Error in Response for Get Last Transaction - Received posRefId:%@ Error:%@. UnknownCompleted.", [gltResponse getPosRefId], [m error]);
```

**Review by mattgresham — COMMENTED** [review]
Thanks for that Doni! I just have a few small changes

File: Library/SPIClient.m (hunk 4 above)
Comment (ghost) [reply Library/SPIClient.m]: I think I was following JS implementation, let me check this

Comment (ghost) [reply Library/SPIClient.m]: Sorry, my fault

**Review by mattgresham — APPROVED** [review]
Thanks for making those changes. Looks good to me!

**Pull request merged**